pkg/datapath/loader/cache.go

```diff
@@ -288,32 +289,39 @@ func (o *objectCache) fetchOrCompile(ctx context.Context, cfg datapath.EndpointC
 	scopedLog := log.WithField(logfields.BPFHeaderfileHash, hash)
 
 	// Serializes attempts to compile this cfg.
+	// TODO(tb): replace with sync.Once.
```

Comment:
:thinking: interesting idea. As long as one goroutine triggers the compile and the others can have a mechanism to block on that (plus some sort of reasonable timeout so we can get retries from further up the stack), this could be worth exploring. But yeah this should not be necessary for solving the problems at hand in the current PR.

Reply:
Yes, a future compilation API would take this into account.

One extra requirement that was overlooked in the existing 'queue' system: multiple endpoints can hash to the same template, but each of those endpoints have independent contexts. A compilation should not be cancelled until all endpoints that want the same hash are cancelled, or cancelling the primary one will cause endpoint compilations/creations to fail for all of them.

Reply:
Yeah good point. It gets a little more tricky with that constraint, just because new endpoints could show up at any time and attempt to add their own context into the mix. To implement the above you would need to support runtime addition of contexts into the list. Probably end up having a whole other goroutine just to handle the context merging and when to actually signal the cancel to the goroutine performing the compile (otherwise, any of the cancelled contexts could interrupt the compile mid-way but there could be new endpoint contexts show up that allow the operation to continue).